pkg/operator/controller/externaldns/namespace.go

```diff
@@ -0,0 +1,81 @@
+/*
+Copyright 2021.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package externaldnscontroller
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// ensureExternalDNSNamespace ensures that the externaldns namespace exists.
+func (r *reconciler) ensureExternalDNSNamespace(namespace string) (bool, *corev1.Namespace, error) {
+	desired := desiredExternalDNSNamespace(namespace)
+
+	haveNamespace, current, err := r.currentExternalDNSNamespace(namespace)
+	if err != nil {
+		return false, nil, err
+	}
+
+	switch {
+	case !haveNamespace:
+		if err := r.createExternalDNSNamespace(desired); err != nil {
+			return false, nil, err
+		}
+		return r.currentExternalDNSNamespace(namespace)
```

Comment:
`desired` variable already holds the instance of the namespace object, no need to get it from API again

Reply:
Actually, some other controllers might add annotations to the namespace, so it's worth fetching it again IMO.

Reply:
This is the pattern we use in the ingress and DNS operators :)

Reply:
Speaking of which, ive added the necessary change to use a non-caching client to this PR. Thanks again!

Should we design the externalDNS operator to (safely) use the default split client, or can we follow suit with the DNS and Ingress operators here?

Reply:
My preference leans towards the approach used in the DNS/Ingress operators simply because it's a pattern I am comfortable with.

Reply:
As discussed in Slack, for starters let's use the not caching client to be sure of the coherent sequential writes/reads.